tests/digraph/test_substitute_node_with_subgraph.py

```diff
@@ -0,0 +1,185 @@
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import unittest
+import retworkx
+
+
+class TestSubstitute(unittest.TestCase):
+    def setUp(self):
+        super().setUp()
+        self.graph = retworkx.generators.directed_path_graph(5)
+
+    def test_empty_replacement(self):
+        in_graph = retworkx.PyDiGraph()
+        res = self.graph.substitute_node_with_subgraph(
+            2, in_graph, lambda _, __, ___: None
+        )
+        self.assertEqual(res, {})
+        self.assertEqual([(0, 1), (3, 4)], self.graph.edge_list())
+
+    def test_single_node(self):
+        in_graph = retworkx.PyDiGraph()
+        in_graph.add_node(0)
+        in_graph.add_child(0, 1, "edge")
+        res = self.graph.substitute_node_with_subgraph(
+            2, in_graph, lambda _, __, ___: 0
+        )
+        self.assertEqual(
+            [(0, 1), (3, 4), (5, 6), (1, 5), (5, 3)], self.graph.edge_list()
+        )
+        self.assertEqual("edge", self.graph.get_edge_data(5, 6))
+        self.assertEqual(res, {0: 5, 1: 6})
+
+    def test_node_filter(self):
+        in_graph = retworkx.PyDiGraph()
+        in_graph.add_node(0)
+        in_graph.add_child(0, 1, "edge")
+        res = self.graph.substitute_node_with_subgraph(
+            2,
+            in_graph,
+            lambda _, __, ___: 0,
+            node_filter=lambda node: node == 0,
+        )
+        self.assertEqual(
+            [(0, 1), (3, 4), (1, 5), (5, 3)], self.graph.edge_list()
+        )
+        self.assertEqual(res, {0: 5})
+
+    def test_edge_weight_modifier(self):
+        in_graph = retworkx.PyDiGraph()
+        in_graph.add_node(0)
+        in_graph.add_child(0, 1, "edge")
+        res = self.graph.substitute_node_with_subgraph(
+            2,
+            in_graph,
+            lambda _, __, ___: 0,
+            edge_weight_map=lambda edge: edge + "-migrated",
+        )
+        self.assertEqual(
+            [(0, 1), (3, 4), (5, 6), (1, 5), (5, 3)], self.graph.edge_list()
+        )
+        self.assertEqual("edge-migrated", self.graph.get_edge_data(5, 6))
+        self.assertEqual(res, {0: 5, 1: 6})
+
+    def test_none_mapping(self):
+        in_graph = retworkx.PyDiGraph()
+        in_graph.add_node(0)
+        in_graph.add_child(0, 1, "edge")
+        res = self.graph.substitute_node_with_subgraph(
+            2, in_graph, lambda _, __, ___: None
+        )
+        self.assertEqual([(0, 1), (3, 4), (5, 6)], self.graph.edge_list())
+        self.assertEqual(res, {0: 5, 1: 6})
```

Comment:
Testing this locally. I got an issue similar to #347 which sometimes the output is `{0: 5, 1: 6}` and `{1: 6, 0: 5}`. Asserting equal to the dictionary doesn't actually catch this difference.
Does order matter for this case?

Reply:
I don't think it matters too much here, because it's a mapping of nodes in the input subgraph to the equivalent node after they're added to the graph. But I will add it to the docstring of the function that it's unordered to make it clear (and so there are no surprises).

Reply:
I added the documentation for this in https://github.com/Qiskit/retworkx/pull/312/commits/80d12164edbeee9c12917125be9b5d6573de1db8